src/libopensc/pkcs15-nqApplet.c

```diff
@@ -0,0 +1,235 @@
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int add_nqapplet_private_key(sc_pkcs15_card_t *p15card, const char *id, int reference, const char *name, const char *pin_id, unsigned int usage)
+{
+	int rv;
+	struct sc_pkcs15_prkey_info prkey_info;
+	struct sc_pkcs15_object prkey_obj;
+	sc_card_t *card = p15card->card;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	memset(&prkey_info, 0, sizeof(prkey_info));
+	memset(&prkey_obj, 0, sizeof(prkey_obj));
+
+	sc_pkcs15_format_id(id, &prkey_info.id);
+	prkey_info.usage = usage;
+	prkey_info.native = 1;
+	prkey_info.key_reference = reference;
+	prkey_info.modulus_length = 3072;
+
+	strlcpy(prkey_obj.label, name, sizeof(prkey_obj.label));
+	prkey_obj.flags = SC_PKCS15_CO_FLAG_PRIVATE;
+	sc_pkcs15_format_id(pin_id, &prkey_obj.auth_id);
+
+	rv = sc_pkcs15emu_add_rsa_prkey(p15card, &prkey_obj, &prkey_info);
+	LOG_TEST_RET(card->ctx, rv, "sc_pkcs15emu_add_rsa_prkey failed");
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int add_nqapplet_objects(sc_pkcs15_card_t *p15card)
+{
+	int rv;
+	sc_card_t *card = p15card->card;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	// 1) User PIN
+	rv = add_nqapplet_pin(p15card, "1", 0x01);
+	LOG_TEST_RET(card->ctx, rv, "Failed to add PIN 1");
+
+	// 2.1) C.CH.Auth
+	rv = add_nqapplet_certificate(p15card, "1", "C.CH.Auth", 0x00);
+	LOG_TEST_RET(card->ctx, rv, "Failed to add Auth. certificate");
+
+	// 2.2) PrK.CH.Auth
+	rv = add_nqapplet_private_key(p15card, "1", 0x01, "PrK.CH.Auth", "1", SC_PKCS15_PRKEY_USAGE_SIGN | SC_PKCS15_PRKEY_USAGE_DECRYPT);
+	LOG_TEST_RET(card->ctx, rv, "Failed to add Auth. private key");
+
+	// 3.1) C.CH.Encr
+	rv = add_nqapplet_certificate(p15card, "2", "C.CH.Encr", 0x01);
+	LOG_TEST_RET(card->ctx, rv, "Failed to add Encr. certificate");
+
+	// 3.2) PrK.CH.Encr
+	rv = add_nqapplet_private_key(p15card, "2", 0x02, "PrK.CH.Encr", "1", SC_PKCS15_PRKEY_USAGE_DECRYPT);
+	LOG_TEST_RET(card->ctx, rv, "Failed to add Encr. private key");
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+int sc_pkcs15emu_nqapplet_init_ex(sc_pkcs15_card_t *p15card, struct sc_aid *aid)
+{
+	int rv = SC_ERROR_WRONG_CARD;
+	sc_context_t *ctx;
+	sc_card_t *card;
+
+	if (!p15card || !p15card->card || !p15card->card->ctx)
+	{
+		return SC_ERROR_INVALID_ARGUMENTS;
+	}
+
+	card = p15card->card;
+	ctx = card->ctx;
+
+	SC_FUNC_CALLED(ctx, SC_LOG_DEBUG_NORMAL);
+
+	if (card->type != SC_CARD_TYPE_NQ_APPLET)
+	{
+		sc_log(p15card->card->ctx, "Unsupported card type: %d", card->type);
+		return SC_ERROR_WRONG_CARD;
+	}
+
+	rv = add_nqapplet_objects(p15card);
+	LOG_TEST_RET(ctx, rv, "Failed to add PKCS15");
+
+	if (aid != NULL)
+	{
+		struct sc_file *file = sc_file_new();
+		if (file != NULL)
+		{
+			/* PKCS11 depends on the file_app object, provide MF */
+			sc_format_path("3f00", &file->path);
+			sc_file_free(p15card->file_app);
+			p15card->file_app = file;
+		}
+	}
+
+	if (p15card->tokeninfo != NULL)
+	{
+		sc_pkcs15_free_tokeninfo(p15card->tokeninfo);
+	}
```

Comment:
The check for NULL is not needed. It is already checked inside of the `sc_pkcs15_free_tokeninfo()` function.

Reply:
Removed it.